lessons/00-intro/best-practices/best_practices_guide.md

```diff
@@ -0,0 +1,260 @@
+python                    3.8.8           h4e93d89_0_cpython    conda-forge
+```
+
+You should not forget to add this new dependence to `environment.yml`. To get the best of both words, our recommended way is:
+
+- All requirements need to be exported in a `requirements.txt` as explained in the `virtualenv` section above
+- Use `environment.yml` to specify your environement name, your python and pip version and get packages dependencies directly from `requirements.txt`
+
+```
+dependencies: # Versions should be taken care of
+  - python=3.10
+  - pip=23.2.1
+  - pip:
+    - -r requirements.txt
+```
+
+### Conda VS Pip
+
+*TL;DR*: use conda.
+
+These two resources are worth reading:
+
+- Stack Overflow: [What is the difference between Pip and Conda](https://stackoverflow.com/questions/20994716/what-is-the-difference-between-pip-and-conda)
+- Anaconda blog: [Understanding the difference between Conda and Pip](https://www.anaconda.com/blog/understanding-conda-and-pip)
+
+### Demo
+
+- Let's create a new environment to run our hello world project using the `install_env.sh` script.
+
+## Locally check your code
+
+You now have a functioning local environment where you can develop python code. Still, it is common for us to make silly mistakes while coding. Over time, these mistakes can compound and add up to a large unexpected failure or just a general decline in code quality. We can employ simple and effective tools to help us overcome these small mistakes when developing.
+
+### Linting
+
+*Linters* perform a *static* evaluation of the code to look for bugs and errors. They are useful as they help in identifying potential issues in the code before it is run. As they perform static evaluation, they cannot catch any RunTime errors. However, a linter can check for syntax errors, type mismatches, and code smells. The most used linters in Python include:
+
+- [Pylint](https://pypi.org/project/pylint/)
+- [Pyflakes](https://pypi.org/project/pyflakes/)
+- [Flake8](https://flake8.pycqa.org/en/latest/)
+- [Ruff](https://github.com/astral-sh/ruff)
+
+Each linter has its own way of functioning and caveats, so choosing which one might be a matter of just sticking to what your pairs already use or looking for a specific feature. In this repo, we use `flake8` whose configuration can be found in `./.flake8`
+
+### Demo
+
+- Lint the `best-practices` folder using flake8
+
+### Automatic Formatting
+
+Good formatting is essetial to ensure that your code is readable and comprehensible. However, it is probably the area of coding in which it is the easiest to make a mistake. That is why we generally use tools that format the code automatically, ensuring that we are always PEP compliant. The most used tool for this end is [Black](https://github.com/psf/black). We also often use a tool to organize imports in a logical manner called [isort](https://pycqa.github.io/isort/). The configuration for both tools can be found in the `./pyproject.toml` file.
```

Comment:
Add nbstripout ? bandit ? safety ? no implemented in the project but mention them so that they know it exists